File: Aoc.Fsharp/Solutions/Day03.fs
module Aoc.Solutions.Day03

open Aoc.Runner

let toInt number =
    let str =
        number
        |> List.map (fun x -> if x then '1' else '0')
        |> fun x -> (System.String.Concat(Array.ofList(x)))


    System.Convert.ToInt32(str, 2)

let parse (input: string) =
    input.Split('\n')
    |> List.ofSeq
    |> List.map (fun x ->
        x
        |> Seq.map (fun c ->
            if c = '1' then true else false)
        |> List.ofSeq)

let mostCommon (nums: list<list<bool>>) i =
    let numsInPos =
        nums
        |> List.map (fun n ->
            n[i]
        )
    let num1 =
        (numsInPos
        |> List.filter (fun x -> x)).Length
    num1 >= (nums.Length + 1) / 2

let leastCommon (nums: list<list<bool>>) i =
    let numsInPos =
        nums
        |> List.map (fun n ->
            n[i]
        )
    (numsInPos
    |> List.filter (fun x -> x)).Length < (nums.Length + 1) / 2

type Day03() =
    inherit Day()

    override _.SolveA input =
        let inp = parse input
        let len = inp[0].Length
        let solve fn =
            seq {
                for i in 0 .. (len - 1) ->
                    fn inp i
            } |> List.ofSeq
        let most = toInt (solve mostCommon)
        let least = toInt (solve leastCommon)
        (most * least) |> string


    override _.SolveB input =
        let inp = parse input
        let len = inp[0].Length
        let solve fn =
            seq {
                let mutable curNums = inp
                for i in 0 .. (len - 1) do
                    let num = fn curNums i
                    curNums <-
                        curNums
                        |> Seq.filter (fun n -> n[i] = num)
                        |> Seq.map List.ofSeq
                        |> List.ofSeq
                    if curNums.Length = 1 then
                        let num = curNums[0]
                        yield num
            }
            |> Seq.exactlyOne
        let a = solve mostCommon |> toInt
        let b = solve leastCommon |> toInt
        (a * b) |> string

    override this.Tests =
        [
            new Test(
                "B",
               @"00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010",
            "230",
            fun x -> this.SolveB(x)
            )
        ]
        |> ResizeArray

File: Aoc.Csharp/Solutions/Day03.cs
using SharpResult.FunctionalExtensions;

namespace Aoc.Solutions;
public class Day03 : Day
{
    static char[] ParseLine(string line) => line.ToCharArray();
    static List<char[]> Parse(string input) => input.Split('\n').Select(ParseLine).ToList();

    public override string SolveA(string input)
    {
        var lines = Parse(input);
        var len = lines.First().Length;
        var gamma = new char[len];
        var eps = new char[len];
        foreach (var i in Enumerable.Range(0, len))
        {
            var col = lines.Select(l => l[i]).ToList();
            gamma[i] = col.Mode();
            eps[i] = gamma[i] == '1' ? '0' : '1';
        }
        var res = Convert.ToInt32(new string(gamma), 2) * Convert.ToInt32(new string(eps), 2);
        return res.ToString();
    }

    public override string SolveB(string input)
    {
        var lines = Parse(input);
        var len = lines.First().Length;

        var GetBits = (IEnumerable<char[]> curLines, int i) => curLines.Select(l => l[i]).ToList();

        var solve = (List<char> bits, Func<List<char>, List<bool>> criteria, List<char[]> intLines) =>
        {
            var output = criteria(bits);
            var res =  intLines.Zip(output).Where(z => z.Second == true).Select(z => z.First);
            return res;
        };

        var oxy = (List<char> bits) =>
        {
            var mostCommon = bits.Mode(def: '1');
            return bits.Select(b => b == mostCommon).ToList();
        };

        var co2 = (List<char> bits) =>
        {
            var mostCommon = bits.Mode(def: '1');
            var leastCommon = mostCommon == '1' ? '0' : '1';
            return bits.Select(b => b == leastCommon).ToList();
        };

        int? oxyRating = null;
        var oxyLines = lines;
        foreach (var i in Enumerable.Range(0, len))
        {
            oxyLines = solve(GetBits(oxyLines, i), oxy, oxyLines).ToList();
            if (oxyLines.Count == 1)
            {
                oxyRating = Convert.ToInt32(new string(oxyLines.First()), 2);
                break;
            }
        }

        int? co2Rating = null;
        var co2Lines = lines;
        foreach (var i in Enumerable.Range(0, len))
        {
            co2Lines = solve(GetBits(co2Lines, i), co2, co2Lines).ToList();
            if (co2Lines.Count == 1)
            {
                co2Rating = Convert.ToInt32(new string(co2Lines.First()), 2);
                break;
            }
        }
        return (co2Rating!.Value * oxyRating!.Value).ToString();
    }

    public Day03()
    {
        Tests = new()
        {
            new("A",
                @"00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010",
            "198",
            SolveA
            ),
            new("B",
                @"00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010",
            "230",
            SolveB
            )
        };
    }
}

static class LinqExt
{
    public static T Mode<T>(this IEnumerable<T> @this, T def = default!)
    {
        var counts = @this.GroupBy(t => t);
        var maxCount = counts.MaxBy(m => m.Count())!.Count();
        var withMaxCount = counts.Where(c => c.Count() == maxCount);
        if (withMaxCount.Count() == 1)
        {
            var mode = withMaxCount.First().Key;
            return mode;
        }
        else
        {
            return def;
        }
    }

    public static string Join<T>(this IEnumerable<T> t, string sep = ", ") => string.Join(sep, t.Select(x => x?.ToString()));
}